### DISCUSSION TITLE
Cannot resolve keyword 'tenant' into field when opening Tenant page

### DISCUSSION THREAD

**Discussion by gfhoffmann** [netbox-community/netbox#16956] [Help Wanted!]
Hello guys!

I'm updating my Netbox from 3.4.7 to 3.7.8 in my local environment and I'm having a problem.

I have several custom plugins, and one of them have models that have FK to the Netbox Tenant model.

After the update the Tenant page is returning me this error (I removed this specific custom plugin and everything is working as expected, so I was able to narrow down the problem to this specific plugin):

```
2024-07-22 17:23:21,755 ERROR [log.py:241] Internal Server Error: /tenancy/tenants/919/
Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/views/generic/base.py", line 26, in dispatch
    return super().dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/utilities/views.py", line 106, in dispatch
    return super().dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/netbox/views/generic/object_views.py", line 77, in get
    **self.get_extra_context(request, instance),
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/netbox/tenancy/views.py", line 129, in get_extra_context
    related_models = [
                     ^
  File "/opt/netbox/netbox/tenancy/views.py", line 130, in <listcomp>
    (model.objects.restrict(request.user, 'view').filter(tenant=instance), f'{field}_id')
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/query.py", line 1436, in filter
    return self._filter_or_exclude(False, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/query.py", line 1454, in _filter_or_exclude
    clone._filter_or_exclude_inplace(negate, args, kwargs)
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/query.py", line 1461, in _filter_or_exclude_inplace
    self._query.add_q(Q(*args, **kwargs))
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1546, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1577, in _add_q
    child_clause, needed_inner = self.build_filter(
                                 ^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1427, in build_filter
    lookups, parts, reffed_expression = self.solve_lookup_type(arg, summarize)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1237, in solve_lookup_type
    _, field, _, lookup_parts = self.names_to_path(lookup_splitted, self.get_meta())
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/netbox/venv/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1725, in names_to_path
    raise FieldError(
django.core.exceptions.FieldError: Cannot resolve keyword 'tenant' into field. Choices are: content, created_at, description, id, project, project_id, updated_at, user, user_id
```

My model is like this:

```
class CustomScript(models.Model):
    """User Custom Script"""
    description = models.TextField(null=True)
    content = models.TextField(null=False)
    user = models.ForeignKey(User, on_delete=models.deletion.CASCADE, related_name='custom_scripts')
    project = models.ForeignKey(Tenant, on_delete=models.deletion.CASCADE, related_name='custom_scripts')
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)
```

Prior to that error I was having:
```
<class 'AttributeError'>

'Manager' object has no attribute 'restrict'
```

But then I added `objects = RestrictedQuerySet.as_manager()` to the model and the error was solved, but then the FieldError showed up.

I changed the name of my `project` field to `tenant`, ran the migration and the error stopped. But changing the model name will be a big refactor on my integrations, so any help here will be appreciated.

Am I missing something here or I really need to change the name of my field? If thats the case, why I need to do this? Is there anything else that I can do to avoid changing the name of the field?

Thanks!

EDIT:

I understood why its happening, in the `TenantView` we have the `get_extra_context()` method that is getting all related models to Tenant.

```
related_models = [
            (model.objects.restrict(request.user, 'view').filter(tenant=instance), f'{field}_id')
            for model, field in get_related_models(Tenant)
        ]
```

When the `for` reaches my custom plugin model, it is breaking in the query: `model.objects.restrict(request.user, 'view').filter(tenant=instance)` because the first part of the query `model.objects.restrict(request.user, 'view')` is returning a RestrictedQuerySet of my CustomScript objects, and when it tries to `filter(tenant=instance)` my model don't have tenant as a field.

But even with that I'm still looking to see if anyone knows something that can help to avoid changing the name of the field. Thanks!

**Comment by candlerb** [comment]
> get_related_models(Tenant)

This comes from netbox/utilities/relations.py

Calling this function, it gets a list of all models which have a FK relationship to Tenant - and it even returns the name of the field which contains that relationship, i.e. it's a list of tuples of `(model, field_name)`

Then deep inside Netbox tenancy views you found a buried, hard-coded assumption that any model which has a FK relationship to Tenant maintains that via an attribute called `tenant`.

However, that code does not exist in Netbox v4.0.7, and it looks like it was completely rewritten in around v4.0.6:

```
commit 5353f837108ee5db11537bc95dde3ce9bc4a042c
Author: Alexander Haase <ahaase@alexhaase.de>
Date:   Wed Jun 12 15:46:41 2024 +0200

    15794 Make "related objects" dynamic (#15876)

    * Closes #15794: Make "related objects" dynamic

    Instead of hardcoding relationships between models for the detail view,
    they are now dynamically generated.

...

 @register_model_view(Tenant)
-class TenantView(generic.ObjectView):
+class TenantView(GetRelatedModelsMixin, generic.ObjectView):
     queryset = Tenant.objects.all()

     def get_extra_context(self, request, instance):
-        related_models = [
-            (model.objects.restrict(request.user, 'view').filter(tenant=instance), f'{field}_id')
-            for model, field in get_related_models(Tenant)
-        ]
-
         return {
-            'related_models': related_models,
+            'related_models': self.get_related_models(request, instance),
         }
```

The new filtering logic from GetRelatedModelsMixin is [here](https://github.com/netbox-community/netbox/blob/v4.0.7/netbox/utilities/views.py#L174-L184).

So you could try updating to latest Netbox v4.0.7 and see if the problem goes away. (I've not tested it myself).

If you can't do that, I think you'll have to go with the flow and change your relationship name to `tenant`.  Even if this is a bug in Netbox, you won't get a bugfix backported to v3.7.8.

**Reply by gfhoffmann** [reply]
Thank you so much for your reply @candlerb 

It is good to know that this was rewritten in the Netbox newer version. I will definitely try it on my local environment, but I will not be able to update Netbox on my prod env yet because we use netbox-secrets as a plugin, and they updated to support Netbox v4 just a week ago and I'm seeing some bugs, so I will wait a little bit more to get a stable version of the plugin.

I don't know when I will be able to reproduce locally to see if the problem was fixed, but when I do I will come back with the feedback!

Thanks again!
